### PR TITLE
cmd/juju: add support for deploying non-public charms.

### PR DESCRIPTION
**Pull request by rogpeppe**
We make AddCharm return an Unauthorized error when appropriate
so that the client code can gather the appropriate macaroon and
pass it to the environment.

(Review request: http://reviews.vapour.ws/r/1401/)

### CODE REVIEW

File: cmd/juju/common.go [hunk 1]
```diff
@@ -136,9 +157,105 @@ func resolveCharmURL(curlStr string, csParams charmrepo.NewCharmStoreParams, rep
...
+// non-public charm deployments. It is defined as a variable so it can
+// be changed for testing purposes.
+var newCharmStoreClient = func() (*csClient, error) {
+	jar, err := cookiejar.New(&cookiejar.Options{
+		PublicSuffixList: publicsuffix.List,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	file := path.Join(utils.Home(), ".go-cookies")
```
Comment (jrwren) [inline cmd/juju/common.go]: consider

    if file := os.Getenv("JUJU_COOKIEFILE"); file != "" {
         return file
    }

It is used elsewhere.

**Comment by rogpeppe** [comment]
$$merge$$

**Comment by rogpeppe** [comment]
$$merge$$

**Comment by rogpeppe** [comment]
$$merge$$

**Comment by rogpeppe** [comment]
$$merge$$

**Pull request merged**